### PR TITLE
fix(cli): harden linear/kata tool outputs against context overruns

### PR DESCRIPTION
**Pull request by gannonh**
## Summary
- add a shared bounded-output helper for Linear/Kata tool rendering
- harden all `linear_*` and `kata_*` tools to avoid large raw payload dumps
- switch broad document/issue inventories to compact summary output with explicit paging/continuation guidance
- add regression coverage for the real overflow paths in Symphony/Kata runs
- update CLI smoke assertions to match current RPC/MCP runtime wiring

## Details
This PR hardens the Linear extension so a single tool call cannot flood agent context with oversized payloads.

Key changes:
- content-bearing reads now use compact metadata plus paged body output
- inventory/list tools now emit compact inventories instead of full content/description payloads
- mutation tools now emit compact summaries instead of echoing full updated objects
- the tool family is audited via `LINEAR_TOOL_STRATEGIES`
- broad document and issue retrieval flows now have explicit regression coverage

## Verification
- `npx vitest run src/resources/extensions/linear/tests/linear-tools.vitest.test.ts`
- `bun test src/resources/extensions/linear/tests/document-operations.test.ts`
- `npx vitest run src/resources/extensions/linear/tests/tool-output.vitest.test.ts src/resources/extensions/linear/tests/linear-client.vitest.test.ts src/resources/extensions/linear/tests/linear-tools.vitest.test.ts`
- `npm run typecheck`
- `npm test`

## Notes
- the exact Node test-runner command for `document-operations.test.ts` still requires globals that are not present in this environment, so that test was verified truthfully with `bun test`
- this work is aimed at preventing Symphony/CLI RPC runs from failing due to sudden Linear/Kata tool output floods

<h3>Greptile Summary</h3>

This PR hardens Linear/Kata tool outputs to prevent context overruns by introducing a `renderPagedTextField` / `renderPagedInventory` / `renderMutationSummary` / `renderCompactRead` helper layer and switching all inventory and mutation tools to compact output. It also adds `listIssueSummaries` and `listDocumentSummaries` queries to `LinearClient` (omitting description/relations), and concurrently removes legacy Desktop model-persistence wiring in favour of reading the model from the CLI runtime on demand.

<h3>Confidence Score: 5/5</h3>

Safe to merge — no P0/P1 issues found; both findings are non-blocking style suggestions.

All tool hardening logic is correct and well-tested. The bounded-output helpers have byte-level tests. The Desktop model-persistence removal is clean. The two P2 findings (dead-code duck-typing fallback, bun test not migrated) do not affect runtime behavior or correctness.

apps/cli/src/resources/extensions/linear/linear-tools.ts — dead-code duck-typing guard in linear_list_documents and kata_list_documents

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| apps/cli/src/resources/extensions/linear/tool-output.ts | New shared output helpers (renderPagedTextField, renderPagedInventory, renderMutationSummary, renderCompactRead, renderInventoryResult, renderErrorSummary) with correct byte-bounding and 1-indexed pagination logic; well tested. |
| apps/cli/src/resources/extensions/linear/linear-tools.ts | All 30+ tools hardened to compact output; linear_list_documents and kata_list_documents use an unnecessary duck-typing guard that creates a dead-code fallback path. |
| apps/cli/src/resources/extensions/linear/linear-client.ts | Adds listIssueSummaries (ISSUE_SUMMARY_FIELDS, no description/relations) and listDocumentSummaries (DOCUMENT_SUMMARY_FIELDS, no content) plus shared filter builders; clean and correct. |
| apps/cli/src/resources/extensions/linear/linear-types.ts | Adds LinearIssueSummary, LinearDocumentSummary, and DocumentListFilter types; structurally sound. |
| apps/cli/src/resources/extensions/linear/tests/tool-output.vitest.test.ts | New Vitest test file covering all four output helpers including byte-bounded truncation and pagination edge cases. |
| apps/desktop/src/main/pi-agent-bridge.ts | Adds refreshSelectedModelFromRuntime (best-effort get_state sync, wrapped in try/catch); called once on first getAvailableModels invocation when selectedModel is null. |
| apps/desktop/src/main/index.ts | Replaces loadPersistedModel / persistSelectedModel with removeLegacyModelSetting; correctly removes the model key via JSON.stringify undefined-drop semantics. |
| apps/cli/src/tests/app-smoke.test.ts | Two assertions updated (extensionFlagValues, runRpcMode(runtime)) to match current CLI API; file remains a bun test rather than being migrated to Vitest per the project policy. |

</details>

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Agent
    participant Tool as linear_list_issues / kata_list_slices
    participant Client as LinearClient
    participant Render as tool-output.ts

    Note over Agent,Render: Before PR — full payload path
    Agent->>Tool: call (no offset/limit)
    Tool->>Client: listIssues(filter)
    Client-->>Tool: LinearIssue[] (with description, relations, children)
    Tool-->>Agent: JSON.stringify(issues) — unbounded

    Note over Agent,Render: After PR — compact inventory path
    Agent->>Tool: call (optional offset/limit)
    Tool->>Client: listIssueSummaries(filter)
    Client-->>Tool: LinearIssueSummary[] (no description/relations)
    Tool->>Render: renderInventoryResult(items, offset, limit)
    Render-->>Tool: paged compact text + continuation guidance
    Tool-->>Agent: bounded output <= HARDENED_MAX_BYTES

    Note over Agent,Render: Content read path (now paged)
    Agent->>Tool: linear_get_issue(id, offset, limit)
    Tool->>Client: getIssue(id)
    Client-->>Tool: LinearIssue (full)
    Tool->>Render: renderCompactRead(metadata, body, offset, limit)
    Render-->>Tool: metadata + paged description
    Tool-->>Agent: bounded output
```

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: apps/cli/src/resources/extensions/linear/linear-tools.ts
Line: 1034-1044

Comment:
**Dead-code fallback via duck-typing guard**

`listDocumentSummaries` is a concrete method on `LinearClient` (defined in `linear-client.ts`), so `typeof summaryClient.listDocumentSummaries === "function"` is always `true` at runtime. The false-branch `summaryClient.listDocuments?.(...)  ?? []` is unreachable production code. The same pattern is duplicated at lines 1568–1578 in `kata_list_documents`.

The simpler, fully type-safe call is just:
```suggestion
        const docsQuery = {
          ...(params.projectId !== undefined ? { projectId: params.projectId } : {}),
          ...(params.first !== undefined ? { first: params.first } : {}),
        };
        const docs = await client.listDocumentSummaries(docsQuery);
```
The same simplification applies to the `kata_list_documents` tool. If supporting test mocks that omit `listDocumentSummaries` is the goal, add it to the mock client in `registerLinearToolsForTest` instead of guarding at the call site.

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: apps/cli/src/tests/app-smoke.test.ts
Line: 543-590

Comment:
**Bun test file modified but not migrated to Vitest**

Per the project testing policy (`apps/cli/CLAUDE.md`): "Vitest is the target runner. Existing bun test files must be migrated to Vitest when their source code is touched." This file (`app-smoke.test.ts`) was changed in this PR to update assertions that reflect current `cli.ts` API names, meaning the underlying source was touched. The file still uses `import assert from "node:assert/strict"` with global `test()` rather than `import { it, expect } from "vitest"`.

Migration is mechanical: rename to `app-smoke.vitest.test.ts`, replace `import assert from "node:assert/strict"` with Vitest's `expect` API.

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Reviews (1): Last reviewed commit: ["test(linear): add overflow regression co..."](https://github.com/gannonh/kata/commit/e1e519ee2cde5c1019c47974cebc80989f20f3f6) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=28141194)</sub>

> Greptile also left **2 inline comments** on this PR.

## Summary by CodeRabbit

* **New Features**
  * Compact summary endpoints for issues and documents with paged listings and slimmer fields
  * New hardened renderers for paged text, inventories, mutation summaries, and error summaries

* **Improvements**
  * Consistent, size‑aware tool output (offset/limit paging, omission guidance, truncation)
  * Desktop: removed legacy persisted model and improved runtime model sync and diagnostics

* **Tests**
  * Expanded unit and end‑to‑end tests covering rendering, paging, tool strategies, and smoke flows

### CODE REVIEW

File: apps/cli/src/resources/extensions/linear/tool-output.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  if (truncation.truncated && truncation.truncatedBy === "bytes") {
+    text += `\n[Truncated to ${formatSize(maxBytes)} while preserving full lines.]`;
+  }
+
+  return truncateHead(text, {
+    maxLines: limit + 4,
+    maxBytes,
+  }).content;
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/linear/tool-output.ts]: Fixed. Refactored to build the footer first, then fit the body within the remaining byte budget so continuation instructions are always preserved.

File: apps/cli/src/resources/extensions/linear/tool-output.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  if (opts.omittedFieldsNote) {
+    text += `\n${opts.omittedFieldsNote}`;
+  }
+
+  const truncation = truncateHead(text, {
+    maxLines: DEFAULT_MAX_LINES,
+    maxBytes: HARDENED_MAX_BYTES,
+  });
+
+  return truncation.content;
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/linear/tool-output.ts]: Fixed. Adopted the suggested approach: build suffix first, then progressively reduce shown items until the candidate fits within the hard limit. Footer and omitted-fields notes are always preserved.

File: apps/cli/src/resources/extensions/linear/tool-output.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  maxBytes?: number;
+}): string {
+  const rawBody = opts.body?.trimEnd() ?? "";
+  if (!rawBody) return opts.emptyMessage;
+
+  const offset = opts.offset ?? 1;
+  const limit = opts.limit ?? DEFAULT_TEXT_PAGE_LIMIT;
+  const maxBytes = opts.maxBytes ?? HARDENED_MAX_BYTES;
+  assertOneIndexedOffset(offset);
+  assertPositiveLimit(limit);
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/linear/tool-output.ts:39]: Fixed. Both renderPagedTextField and renderPagedInventory now clamp limit to their respective defaults (DEFAULT_TEXT_PAGE_LIMIT / DEFAULT_LIST_PAGE_LIMIT).

File: apps/cli/src/resources/extensions/linear/tool-output.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+  const startIndex = offset - 1;
+  const page = opts.items.slice(startIndex, startIndex + limit);
+  const lines = page.map((item, pageIndex) => opts.renderItem(item, startIndex + pageIndex + 1));
+  let text = lines.join("\n");
+  const shownStart = offset;
+  const shownEnd = offset + page.length - 1;
+
+  if (shownEnd < opts.items.length) {
+    text += `\n\n[Showing items ${shownStart}-${shownEnd} of ${opts.items.length}. Use offset=${shownEnd + 1} to continue.]`;
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/linear/tool-output.ts]: Fixed. Dropped "of N" wording from renderPagedInventory since items.length may be a capped fetch. The text field pager keeps "of N" since line counts are always exact.

File: apps/cli/src/resources/extensions/linear/linear-client.ts [hunk 5]
```diff
@@ -575,6 +578,65 @@ export class LinearClient {
...
+      name
+      type
+      color
+      position
+    }
+    labels {
+      nodes {
+        id
+        name
+        color
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/linear/linear-client.ts:601]: Fixed. Added isGroup to ISSUE_SUMMARY_FIELDS so the query matches the LinearLabel type contract.

File: apps/cli/src/resources/extensions/linear/linear-tools.ts [hunk 6]
```diff
@@ -51,28 +57,79 @@ export {
...
+function okText(text: string) {
   return {
-    content: [{ type: "text" as const, text: JSON.stringify(data, null, 2) }],
+    content: [{ type: "text" as const, text }],
   };
 }
 
+function ok(data: unknown) {
+  return okText(JSON.stringify(data, null, 2));
+}
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/linear/linear-tools.ts:118]: The two ok() call sites (lines 1649, 1665) return fixed-size "blocked" payloads well under 1KB. The general path goes through run() which returns strings rendered by bounded helpers. No risk of oversized output from these static objects.

File: apps/cli/src/resources/extensions/linear/linear-tools.ts [hunk 7]
```diff
@@ -313,8 +601,30 @@ export function registerLinearTools(pi: ExtensionAPI, client: LinearClient) {
...
       assigneeId: Type.Optional(Type.String({ description: "Filter by assignee UUID" })),
       first: Type.Optional(Type.Number({ description: "Max results per page (default: 50)" })),
+      offset: Type.Optional(Type.Integer({ minimum: 1, description: "Item number to start from (1-indexed)" })),
+      limit: Type.Optional(Type.Integer({ minimum: 1, description: "Maximum number of items to return" })),
     }),
-    async execute(_id, params) { return run(() => client.listIssues(params)); },
+    async execute(_id, params) {
+      return run(async () => {
+        const { offset, limit, ...filters } = params;
+        const issues = await client.listIssueSummaries(filters);
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/linear/linear-tools.ts:609]: Valid optimization. LinearClient.paginate already caps at maxPages so the fetch is bounded, and the rendering layer enforces the hard byte limit. Pushing offset/limit into the query is a good follow-up but outside this PR scope (output hardening, not query optimization).

File: apps/cli/src/resources/extensions/linear/linear-tools.ts [hunk 8]
```diff
@@ -504,33 +967,96 @@ export function registerLinearTools(pi: ExtensionAPI, client: LinearClient) {
...
+          listDocumentSummaries?: (opts?: { projectId?: string; issueId?: string; title?: string; first?: number }) => Promise<Array<{ id: string; title: string; project?: { id: string; name: string } | null; issue?: { id: string; identifier: string } | null; updatedAt: string }>>;
+          listDocuments?: (opts?: { projectId?: string; issueId?: string; title?: string; first?: number }) => Promise<Array<{ id: string; title: string; project?: { id: string; name: string } | null; issue?: { id: string; identifier: string } | null; updatedAt: string }>>;
+        };
+        const docsQuery = {
+          ...(params.projectId !== undefined ? { projectId: params.projectId } : {}),
+          ...(params.first !== undefined ? { first: params.first } : {}),
+        };
+        const docs = typeof summaryClient.listDocumentSummaries === "function"
+          ? await summaryClient.listDocumentSummaries(docsQuery)
+          : await summaryClient.listDocuments?.(docsQuery) ?? [];
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/linear/linear-tools.ts]: Same reasoning as the issues paging comment. The fetch is already bounded by maxPages, and the rendering layer enforces byte limits. Query-level paging is a valid follow-up optimization.

[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/linear/linear-tools.ts]: Fixed. Removed the duck-typing guards at both linear_list_documents and kata_list_documents. Both now call client.listDocumentSummaries() directly.

File: apps/cli/src/tests/app-smoke.test.ts [hunk 9]
```diff
@@ -583,8 +583,8 @@ test("cli.ts routes --mode rpc through runRpcMode", () => {
     "cli.ts lazily imports runRpcMode",
   );
   assert.ok(
-    cliSrc.includes("await runRpcMode(session)"),
-    "cli.ts invokes runRpcMode(session)",
+    cliSrc.includes("await runRpcMode(runtime)"),
+    "cli.ts invokes runRpcMode(runtime)",
   );
 });
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/tests/app-smoke.test.ts]: Fixed. Migrated app-smoke.test.ts to app-smoke.vitest.test.ts with Vitest describe/it/expect API. Old bun test file deleted.

File: apps/cli/src/resources/extensions/linear/linear-tools.ts [hunk 10]
```diff
@@ -91,10 +148,29 @@ export function registerLinearTools(pi: ExtensionAPI, client: LinearClient) {
...
+          noun: "teams",
+          items: teams,
+          offset: params.offset,
+          limit: params.limit,
+          omittedFieldsNote: "Large fields omitted from list output. Use linear_get_team to inspect one team.",
+          renderItem: (team, index) => [
+            `${index}. ${team.key}: ${team.name}`,
+            `   id: ${team.id}`,
+            `   description: ${team.description || "—"}`,
+          ].join("\n"),
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/linear/linear-tools.ts:169]: Fixed in `43e78bc9`. The compact team and label inventory renderers now omit verbose descriptions so a long description cannot consume the page budget and crowd out the actual rows.

File: apps/cli/src/resources/extensions/linear/tool-output.ts [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  const footer = footerParts.length > 0 ? "\n\n" + footerParts.join("\n") : "";
+  const footerBytes = Buffer.byteLength(footer, "utf8");
+  const bodyBudget = maxBytes - footerBytes;
+
+  const fittedBody = bodyBudget > 0
+    ? truncateHead(truncation.content, { maxLines: limit, maxBytes: bodyBudget }).content
+    : "";
+  return fittedBody + footer;
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/linear/tool-output.ts]: Fixed in `43e78bc9`. `renderPagedTextField()` now recomputes continuation guidance from the final fitted body after footer-budget truncation, including the zero-visible-lines edge case, so the suggested `offset` never skips unseen lines.

File: apps/cli/src/tests/app-smoke.vitest.test.ts [hunk 12]
```diff
@@ -0,0 +1,484 @@
...
+      encoding: "utf-8",
+    });
+    const packInfo = JSON.parse(packOutput);
+    const tarball = packInfo[0].filename;
+    const tarballPath = join(projectRoot, tarball);
+
+    expect(existsSync(tarballPath)).toBe(true);
+
+    try {
+      const contents = execSync(`tar tzf ${tarballPath}`, { encoding: "utf-8" });
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/tests/app-smoke.vitest.test.ts]: Fixed in `43e78bc9`. The smoke tests now use `execFileSync(...)` with argv arrays for `tar` and `npm install`, so temp paths are not shell-interpolated.

File: apps/desktop/package.json [hunk 13]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@kata/desktop",
-  "version": "0.2.1",
+  "version": "0.2.2",
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/package.json:3]: Keeping this intentionally. The Desktop version bump is part of the requested release prep for this PR — the user explicitly asked for patch releases for both CLI and Desktop here, so `apps/desktop/package.json` staying at `0.2.2` is deliberate rather than accidental drift.

File: apps/desktop/src/main/index.ts [hunk 14]
```diff
@@ -222,10 +211,10 @@ async function persistWorkspacePath(workspacePath: string): Promise<void> {
 }
 
 app.whenReady().then(async () => {
+  await removeLegacyModelSetting()
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/main/index.ts]: Fixed in `43e78bc9`. `removeLegacyModelSetting()` is now wrapped as best-effort cleanup during startup: failures are logged with `log.warn(...)` and no longer block app readiness.
